### PR TITLE
Record: Bank QAT + seq4096 + SWA w=256 + QK-Gain 2.5 + PKO — val_bpb 1.1117 (3-seed mean)

### PR DESCRIPTION
**Pull request by Itssshikhar**
## Summary

  **val_bpb: 1.1117** (3-seed mean, std 0.0008) | **~16.1 MB** | 8×H100 SXM, 600s | No TTT

  Built on PR #1019 stack. Key additions: seq4096 training, bank-weight QAT (fake int6 on all 26M F.linear params), QK-Gain 2.5, and partial key offset from NanoGPT speedrun.

  ## Results (8×H100 80GB SXM, 600s, no TTT)

  | Seed | Steps | ms/step | Post-EMA BPB | Roundtrip BPB | **Sliding BPB** | Artifact |
  |------|-------|---------|--------------|---------------|-----------------|----------|
  | 1337 | 7,298 | 82.22 | 1.1201 | 1.1252 | **1.1113** | 16,118,346 |
  | 42 | 7,298 | 81.51 | 1.1198 | 1.1250 | **1.1112** | 16,109,918 |
  | 7 | 7,204 | 83.29 | 1.1213 | 1.1263 | **1.1126** | 16,120,506 |
  | **Mean** | **7,267** | **82.34** | **1.1204** | **1.1255** | **1.1117** | **16,116,257** |

  Delta vs SOTA PR #1019 (1.1147): **-0.0030 BPB**.

  ## Key Techniques

  1. **seq4096 training** — biggest single gain, ~0.009 BPB pre-quant improvement over seq2048
  2. **Bank-weight QAT** — STE fake int6 quantization on all F.linear bank weights (q/k/v/o/up/down) during late training, not just CastedLinear layers
  3. **QK-Gain 2.5** — higher initial gain for QK-normalization (from NanoGPT speedrun)
  4. **Partial Key Offset** — shift stationary RoPE dims of K forward by 1 position for single-layer induction heads
  5. **SWA w=256** — sliding window attention on layers 0-5, full attention on layers 6-10
  6. **Full Hessian GPTQ** — AR self-gen calibration (64 seqs × 2048 tokens), int6 per-row quantization
  7. **LZMA compression** — preset 9

  ## Negative Results

  - Sparse attention gate: helps pre-quant but widens quant gap (+0.0016)
  - Sigmoid logit rescale: widens quant gap
  - Cautious weight decay: catastrophic for quantization
  - Batch scheduling (3 variants): worse
  - MTP (0.2 and 0.02): hurts
  - N-gram hedge mixer: neural model too strong, 100% weight to neural
  - TTT: neutral
  - Re-quantize at different seq_len: model incompatible
  - Hadamard rotation: no effect
  - SpQR analysis: no outliers
  - Mixed int5/int6: too lossy

  ## Reproduction

  ```bash
  SEED=$SEED RUN_ID=bankqat_s$SEED TRAIN_SEQ_LEN=4096 EVAL_SEQ_LEN=4096 \
    SWA_WINDOW_SIZE=256 SWA_FULL_ATTN_LAYERS=5 PARTIAL_KEY_OFFSET=1 \
    BIGRAM_VOCAB_SIZE=3072 BIGRAM_DIM=112 WARMDOWN_ITERS=4000 \
    torchrun --standalone --nproc_per_node=8 train_gpt_swa.py
  ```

  Test Plan

  - 3-seed validation on Modal 8×H100 (seeds 1337, 42, 7)
  - All artifacts under 16,777,216 bytes
  - Training under 600s
  - Full pipeline: train → EMA → GPTQ → compress → decompress → dequantize → sliding eval

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: Bank QAT + seq4096 + SWA w=256 + QK-Gain 2.5 + PKO — val_bpb 1.1117 (3-seed mean)

**BPB:** 1.1117 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `3cca46cbfba0`, file `records/track_10min_16mb/2026-03-20_11L_EfficientPartialXSA_FA3_SWA120/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.02s, dim=512, layers=9, vocab=1024, code=66055 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.02s, dim=512, layers=9, vocab=1024, code=66055 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
